engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterWindowController.mm

```diff
@@ -0,0 +1,197 @@
+  FlutterViewController* c = [[FlutterViewController alloc] initWithEngine:_engine
+                                                                   nibName:nil
+                                                                    bundle:nil];
+
+  NSWindow* window = [[NSWindow alloc] init];
+  // If this is not set there will be double free on window close when
+  // using ARC.
+  [window setReleasedWhenClosed:NO];
+
+  window.contentViewController = c;
+  [window setContentSize:NSMakeSize(request->width, request->height)];
+  window.styleMask = NSWindowStyleMaskResizable | NSWindowStyleMaskTitled |
+                     NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable;
+  [window setIsVisible:YES];
+  [window makeKeyAndOrderFront:nil];
+
+  FlutterWindowOwner* w = [[FlutterWindowOwner alloc] initWithWindow:window
+                                               flutterViewController:c
+                                                     creationRequest:*request];
+  window.delegate = w;
+  [_windows addObject:w];
+
+  return c.viewIdentifier;
+}
+
+- (void)destroyWindow:(NSWindow*)window {
+  FlutterWindowOwner* owner = nil;
+  for (FlutterWindowOwner* o in _windows) {
+    if (o.window == window) {
+      owner = o;
+      break;
+    }
+  }
+  if (owner != nil) {
+    [_windows removeObject:owner];
+    [owner.flutterViewController dispose];
+    owner.window.delegate = nil;
+    [owner.window close];
+  }
+}
+
+@end
+
+// NOLINTBEGIN(google-objc-function-naming)
+
+int64_t FlutterCreateRegularWindow(int64_t engine_id, const FlutterWindowCreationRequest* request) {
+  FlutterEngine* engine = [FlutterEngine engineForIdentifier:engine_id];
+  [engine enableMultiView];
+  return [engine.windowController createRegularWindow:request];
+}
+
+void FlutterDestroyWindow(int64_t engine_id, void* window) {
+  NSWindow* w = (__bridge NSWindow*)window;
+  FlutterEngine* engine = [FlutterEngine engineForIdentifier:engine_id];
+  [engine.windowController destroyWindow:w];
+}
+
+void* FlutterGetWindowHandle(int64_t engine_id, FlutterViewIdentifier view_id) {
+  FlutterEngine* engine = [FlutterEngine engineForIdentifier:engine_id];
+  FlutterViewController* controller = [engine viewControllerForIdentifier:view_id];
+  return (__bridge void*)controller.view.window;
+}
+
+void FlutterGetWindowSize(void* window, FlutterWindowSize* size) {
+  NSWindow* w = (__bridge NSWindow*)window;
+  size->width = w.frame.size.width;
+  size->height = w.frame.size.height;
+}
+
+void FlutterSetWindowSize(void* window, double width, double height) {
+  NSWindow* w = (__bridge NSWindow*)window;
+  [w setContentSize:NSMakeSize(width, height)];
+}
+
+void FlutterSetWindowTitle(void* window, const char* title) {
+  NSWindow* w = (__bridge NSWindow*)window;
+  w.title = [NSString stringWithUTF8String:title];
+}
+
+int64_t FlutterGetWindowState(void* window) {
+  NSWindow* w = (__bridge NSWindow*)window;
+  if (w.isZoomed) {
+    return 1;
```

Comment:
Use an enum for these?

Reply:
Yeah, I'll use the values from `flutter::WindowState`.